dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/CallBase.java

```diff
@@ -0,0 +1,85 @@
+package com.dat3m.dartagnan.program.event.common;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionVisitor;
+import com.dat3m.dartagnan.expression.Type;
+import com.dat3m.dartagnan.expression.type.FunctionType;
+import com.dat3m.dartagnan.program.Function;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.Register.UsageType;
+import com.dat3m.dartagnan.program.event.AbstractEvent;
+import com.dat3m.dartagnan.program.event.CallEvent;
+import com.google.common.base.Preconditions;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public abstract class CallBase extends AbstractEvent implements CallEvent {
+
+    protected FunctionType funcType;
+    protected Expression callTarget;
+    protected List<Expression> arguments;
+
+    protected CallBase(FunctionType funcType, Expression funcPtr, List<Expression> arguments) {
+        final List<Type> paramTypes = funcType.getParameterTypes();
+        Preconditions.checkArgument(
+                (!funcType.isVarArgs() && arguments.size() == paramTypes.size())
+                        || (funcType.isVarArgs() && arguments.size() >= paramTypes.size())
+        );
+        for (int i = 0; i < paramTypes.size(); i++) {
+            Preconditions.checkArgument(arguments.get(i).getType().equals(paramTypes.get(i)));
+        }
+        this.funcType = funcType;
+        this.callTarget = funcPtr;
+        this.arguments = new ArrayList<>(arguments);
+    }
+
+    protected CallBase(CallBase other) {
+        super(other);
+        this.funcType = other.funcType;
+        this.callTarget = other.callTarget;
+        this.arguments = new ArrayList<>(other.arguments);
+    }
+
+    public boolean isDirectCall() { return callTarget instanceof Function; }
```

Comment:
Do you really need this one? I think the current default is doing exactly the same (since `getCallTarget() { return callTarget; }`)